### PR TITLE
Extension of the Worker Class 

### PR DESCRIPTION
**Pull request by jan-janssen**

### CODE REVIEW

File: pyiron_base/job/worker.py [hunk 1]
```diff
@@ -207,3 +207,36 @@ def run_static(self):
...
+        master_id = self.job_id
+        while not finished:
+            df = pr.job_table()
+            df_sub = df[
+                ((df["status"] == "submitted") | (df.status == "running")) &
+                (df["masterid"] == master_id)
+            ]
+            if len(df_sub) == 0:
+                j += 1
+                if j > 10:
```
Comment (liamhuber) [inline pyiron_base/job/worker.py]: If I read right this is a safety feature so you need consecutive blank calls before really considering yourself done. In this case 10 should be a kwarg, and it should probably be in units of seconds and related to the other new kwarg

File: pyiron_base/job/worker.py [hunk 2]
```diff
@@ -207,3 +207,36 @@ def run_static(self):
...
+                    finished = True
+            else:
+                j = 0
+            with open(log_file, 'a') as f:
+                f.write(
+                    str(datetime.today()) + " j: " + str(j) +
+                    "   submitted: " + str(len(df[df.status == "submitted"])) +
```
Comment (liamhuber) [inline pyiron_base/job/worker.py]: Nit: this might read better in loop form over the statuses

File: pyiron_base/job/worker.py [hunk 3]
```diff
@@ -207,3 +207,36 @@ def run_static(self):
...
+                    "   running: " + str(len(df[df.status == "running"])) +
+                    "   finished: " + str(len(df[df.status == "finished"])) +
+                    "   aborted: " + str(len(df[df.status == "aborted"])) + "\n"
+                )
+            time.sleep(60)
```
Comment (liamhuber) [inline pyiron_base/job/worker.py]: 60 should be a kwarg. Default of 60 is of course fine.

**Review by liamhuber — COMMENTED** [review]
Would be nice to offer some flexibility in how patient the user wants to be

**Comment by pmrv** [comment]
Could this be implemented in terms of `Project.wait_for_jobs`? If not then the arguments @liamhuber suggested should match that function.  I'm fine with having this extra function for convenience, but we should avoid duplicated logic or interfaces.

**Comment by jan-janssen** [comment]
> Could this be implemented in terms of `Project.wait_for_jobs`? If not then the arguments @liamhuber suggested should match that function. I'm fine with having this extra function for convenience, but we should avoid duplicated logic or interfaces.

The difference is basically that it searches for jobs based on the `master_id` rather than the path, which allows you to have multiple workers inside the same project and it sets the status of the worker to `collect` once all jobs are finished to stop the worker. So while I foresee in the long run the `worker` jobs just like `pyirontable` being meta jobs the user never actively works with, at the current stage I prefer to have a separate function.

**Comment by liamhuber** [comment]
I see the merit of having such `wait` code directly on the worker master, but Marvin's point about not duplicating things is also a very good one. Can we have our cake and eat it too? 

What if we extract the waiting code, such that the source, times, etc. are all taken as arguments, then exploit this extracted behaviour at both the worker and project levels. The additional behaviour that the worker sets itself to collect should be easy enough to append after.

I'm sitting at 30k ft on this, so forgive me if I miss some impossible architecture obstacle. Seems like it should work though..

**Comment by jan-janssen** [comment]
> I'm sitting at 30k ft on this, so forgive me if I miss some impossible architecture obstacle. Seems like it should work though..

While the functions have a similar effect for the user, their technical implementation is different e.g. filter by master id, write log file and only finish after time out, for the case the tasks for the worker are submitted from a different notebook. So I currently do not see an option to combine them. Internally they both use the `job_table()` to get the data from the database but the filtering and the reaction are different.

**Pull request merged**